odoo/models.py

```diff
@@ -1731,19 +1768,325 @@ def clear_caches(cls):
+            if granularity == 'week':
+                # first_week_day: 0=Monday, 1=Tuesday, ...
+                first_week_day = int(get_lang(self.env).week_start) - 1
+                days_offset = first_week_day and 7 - first_week_day
+                field_expression = f"(date_trunc('week', {field_expression}::timestamp - INTERVAL '-{days_offset} DAY') + INTERVAL '-{days_offset} DAY')"
+            else:
+                field_expression = f"date_trunc('{granularity}', {field_expression}::timestamp)"
+            if field.type == 'date':
+                field_expression += '::date'
+        elif field.type == 'boolean':
+            field_expression = f"COALESCE({field_expression}, FALSE)"
+
+        return field_expression, [fname]
+
+    @api.model
+    def _read_group_having(self, query, having_domain):
+        fnames_used = []
+        having_params = []
+        stack_op = []
+
+        def pop_next_condition():
+            stack_leaf = stack_op.pop()
+            if isinstance(stack_leaf, str):
+                return stack_leaf
+
+            aggragate_spec, ope, value = stack_leaf
+            having_params.append(value)
+            select_term, select_fnames_used = self._read_group_select(query, aggragate_spec)
+            fnames_used.extend(select_fnames_used)
+            return f'{select_term} {ope} %s'
+
+        for leaf in reversed(having_domain):
+            if leaf == '!':
+                condition_1 = pop_next_condition()
+                stack_op.append(f'(NOT {condition_1})')
+            elif leaf in ('&', '|'):
+                condition_1 = pop_next_condition()
+                condition_2 = pop_next_condition()
+
+                op = 'AND' if leaf == '&' else 'OR'
+                stack_op.append(f'({condition_1} {op} {condition_2})')
+            elif isinstance(leaf, (list, tuple)):
+                __, operator, __ = leaf
+                if operator not in ('in', 'not in', '<', '>', '<=', '>=', '=', '!='):
+                    raise ValueError(f"Having clause {leaf!r} only accept ('in', 'not in', '<', '>', '<=', '>=', '=', '!=') operators")
+                stack_op.append(leaf)
+            else:
+                raise ValueError(f"Invalid having clause {leaf!r}. It should be a domain-like value")
+
+        while len(stack_op) > 1:
+            condition_1 = pop_next_condition()
+            condition_2 = pop_next_condition()
+            stack_op.append(f'({condition_1} AND {condition_2})')
+
+        return pop_next_condition() if stack_op else "", having_params, fnames_used
+
+    @api.model
+    def _read_group_orderby(self, query, groupby, order, order_traverse_many2one):
+        orderby_terms = []
+        extra_groupby_terms = []
+        fnames_used = []
+        if not order:
+            return orderby_terms, extra_groupby_terms, fnames_used
+
+        for order_term in order.split(','):
+            order_field, order_orientation = order_term.split()
```

Comment:
this line crash with
```py
self.env['res.partner']._read_group([], ['name'])
```

shouldn't we have one by default ?

Reply:
It was a small regression done this morning. Already fix.